### PR TITLE
PT-2340: support MySQL 8.4

### PR DESCRIPTION
**Pull request by svetasmirnova**
This huge PR adds MySQL 8.4 support to the Percona Toolkit.

- [x] The contributed code is licensed under GPL v2.0
- [x] Contributor Licence Agreement (CLA) is signed
- [x] util/update-modules has been ran
- [x] Documentation updated
- [x] Test suite update

### CODE REVIEW

File: t/lib/bash/collect_mysql_info.sh [hunk 1]
```diff
@@ -163,11 +163,18 @@ cmd_ok \
 # collect_master_logs_status
 
 if [ -n "$(get_var log_bin "$p/mysql-variables")" ]; then
+   mysql_version="$(get_var version "$p/mysql-variables")"
+   source_logs_file="mysql-binary-logs"
+   source_status_file="mysql-binary-log-status"
+   if [ "$mysql_version" '<' "8.1" ]; then
```
Comment (ylacancellera) [inline t/lib/bash/collect_mysql_info.sh:169]: Is that planned for any mariadb version ? 
```
$ if [ "10.4" '<' "8.1" ]; then echo smaller; fi
smaller
```

File: t/pt-archiver/archive_using_channels.t [hunk 2]
```diff
@@ -17,67 +17,71 @@ require "$trunk/bin/pt-archiver";
...
+   server => 'chan_replica1',
+   type   => 'source',
 );
-my $slave1_port = $sb->port_for('chan_slave1');
+my $replica1_port = $sb->port_for('chan_replica1');
 
-$sb->load_file('chan_master1', "sandbox/gtid_on.sql", undef, no_wait => 1);
-$sb->load_file('chan_master2', "sandbox/gtid_on.sql", undef, no_wait => 1);
-$sb->load_file('chan_slave1', "sandbox/slave_channels.sql", undef, no_wait => 1);
+if ( $sandbox_version lt '8.1' ) {
```
Comment (ylacancellera) [inline t/pt-archiver/archive_using_channels.t:48]: commenting again for -lt, this time on perl, just in case. there's MySQL 9 coming, I imagine MySQL 10 would exist at a point.
```
$ perl -e "if ( '10.0' lt '8.1') { print('smaller')}"
smaller
```

File: bin/pt-archiver [hunk 3]
```diff
@@ -7179,14 +7292,14 @@ sub main {
             my $lag_dbh = $lag_server->{'dbh'};
             my $id      = $lag_server->{'id'};
             if ( $lag_dbh ) {
-               my $lag = $ms->get_slave_lag($lag_dbh);
+               my $lag = $ms->get_replica_lag($lag_dbh);
                while ( !defined $lag || $lag > $o->get('max-lag') ) {
                   PTDEBUG && _d("Sleeping: slave lag for server '$id' is", $lag);
```
Comment (ylacancellera) [inline bin/pt-archiver]: Since pt-heartbeat had his schema changed to reflect the source/replica change, does this one should be changed for coherence ?
Or, is it left as is so to keep the current output ?

File: t/pt-heartbeat/multi_update_mode.t [hunk 4]
```diff
@@ -60,175 +61,231 @@ foreach my $port (@ports) {
...
    defined $rows->{12346}->{file} && defined $rows->{12346}->{position},
-   "Slave1 master file and position"
+   "Replica1 source file and position"
 );
 
 ok(
-   $rows->{12346}->{relay_master_log_file} && $rows->{12346}->{exec_master_log_pos},
-   "Slave1 relay_master_log_file and exec_master_log_pos for master"
-);
+   $rows->{12346}->{"relay_source_log_file"} && $rows->{12346}->{"exec_source_log_pos"},
```
Comment (ylacancellera) [inline t/pt-heartbeat/multi_update_mode.t:113]: info: why is this test (and next one) not using `relay_$(source_name}_log_file` like the previous one ? Can it miss an edge case ?
"test" as in `ok(some conditions)` not as in test file

File: bin/pt-archiver (hunk 3 above)
Comment (svetasmirnova) [reply bin/pt-archiver]: Hi! I don't understand your comment. Modified code uses get_replica_lag. What needs to be changed?

File: t/pt-heartbeat/multi_update_mode.t (hunk 4 above)
Comment (svetasmirnova) [reply t/pt-heartbeat/multi_update_mode.t:113]: Because it selects from the test.heartbeat table, and not from SHOW SLAVE/REPLICA STATUS. This table has static column names, not depending on server version. Legacy column names are supported, this is tested by different test: see t/pt-heartbeat/heartbeat_table.t

File: t/pt-archiver/archive_using_channels.t (hunk 2 above)
Comment (svetasmirnova) [reply t/pt-archiver/archive_using_channels.t:48]: Good point! We do not care about MariaDB in our tests atm but we will in future versions. Could you please file a Jira ticket, so I can plan a fix for it?

File: t/lib/bash/collect_mysql_info.sh (hunk 1 above)
Comment (svetasmirnova) [reply t/lib/bash/collect_mysql_info.sh:169]: Good point! We do not care about MariaDB in our tests atm but we will in future versions. Could you please file a Jira ticket, so I can plan a fix for it?

File: bin/pt-archiver (hunk 3 above)
Comment (ylacancellera) [reply bin/pt-archiver]: It is for `_d("Sleeping: slave lag for server '$id' is", $lag);`, on row 7297 and 7299 (on new branch).
7297 is with PTDEBUG, so really not a problem, but the other one is a log to users
"replica lag" would make it more coherent, but that's only a nitpick

File: t/pt-heartbeat/multi_update_mode.t (hunk 4 above)
Comment (ylacancellera) [reply t/pt-heartbeat/multi_update_mode.t:113]: Got it thank you

File: bin/pt-archiver (hunk 3 above)
Comment (svetasmirnova) [reply bin/pt-archiver]: Fixed, please review again

Comment (ylacancellera) [reply bin/pt-archiver]: lgtm for the bin, the tests will need some fixing as well
```
$ grep -rn 'Sleeping:'
t/pt-archiver/replica_lag.t:73:   qr/Sleeping: slave lag for server/,
t/pt-archiver/replica_lag.t:101:   qr/Sleeping: slave lag for server/,
t/pt-archiver/replica_lag.t:129:   qr/Sleeping: slave lag for server/,
t/pt-archiver/replica_lag.t:156:   qr/Sleeping: slave lag for server/,
t/pt-archiver/replica_lag.t:184:   qr/Sleeping: slave lag for server/,
t/pt-archiver/replica_lag.t:212:   qr/Sleeping: slave lag for server/,
t/pt-archiver/replica_lag.t:240:   qr/Sleeping: slave lag for server/,
bin/pt-archiver:7297:                  PTDEBUG && _d("Sleeping: replica lag for server '$id' is", $lag);
bin/pt-archiver:7299:                     _d("Sleeping: replica lag for server '$id' is", $lag);
```

File: t/pt-archiver/archive_using_channels.t (hunk 2 above)
Comment (ylacancellera) [reply t/pt-archiver/archive_using_channels.t:48]: https://perconadev.atlassian.net/browse/PT-2390

File: t/lib/bash/collect_mysql_info.sh (hunk 1 above)
Comment (ylacancellera) [reply t/lib/bash/collect_mysql_info.sh:169]: https://perconadev.atlassian.net/browse/PT-2390

File: lib/MySQLProtocolParser.pm [hunk 5]
```diff
@@ -75,7 +75,7 @@ use constant {
    COM_BINLOG_DUMP         => '12',
    COM_TABLE_DUMP          => '13',
    COM_CONNECT_OUT         => '14',
-   COM_REGISTER_SLAVE      => '15',
+   COM_REGISTER_REPLICA    => '15',
```
Comment (ylacancellera) [inline lib/MySQLProtocolParser.pm]: The packet exists on 8.0 and 8.4, but not 5.7
I don't find any code usage of it in any lib/bin though

File: lib/PerconaTest.pm [hunk 6]
```diff
@@ -78,6 +83,19 @@ eval {
    $sandbox_version = $v if $v;
 };
 
+our $source_name = 'source';
+our $source_status = 'binary log';
+our $source_reset = 'binary logs and gtids';
+our $source_change = 'replication source';
+our $replica_name = 'replica';
+if ( $sandbox_version lt '8.1' || ( $ENV{FORK} || "" eq 'mariadb' ) ) {
```
Comment (ylacancellera) [inline lib/PerconaTest.pm:91]: works in an unintended way (PT-2390). I listed this code path in PT-2390 for review. I'm still commenting since it's a different condition
10.x will return true on `$sandbox_version lt '8.1'`, so the second part on the fork is not used.
```
$ echo $sandbox_version $FORK
10.4 mysql
$ perl -e "if ( $sandbox_version lt '8.1' || ( $ENV{FORK} || '' eq 'mariadb' ) ) { print('works') }"
works
```
As long as mysql 10 does not exist it still does what we want

File: bin/pt-replica-find [hunk 7]
```diff
@@ -0,0 +1,4819 @@
...
+
+# ###########################################################################
+# This is a combination of modules and programs in one -- a runnable module.
+# http://www.perl.com/pub/a/2006/07/13/lightning-articles.html?page=last
+# Or, look it up in the Camel book on pages 642 and 643 in the 3rd edition.
+#
+# Check at the end of this package for the call to main() which actually runs
+# the program.
+# ###########################################################################
+package pt_slave_find;
```
Comment (ylacancellera) [inline bin/pt-replica-find]: can this impact maintenance ? The names are mismatched between the toolname and the package name
If it's safe, good to me. I don't find any direct usage

File: bin/pt-replica-find [hunk 8]
```diff
@@ -0,0 +1,4819 @@
...
+#    master:port
+#    +- slave1:port
+#       +- slave1-1:port
+#    +- slave2:port
+# The levels are thus:
+#    0
+#    +- 1
+#       +- 2
+#    +- 1
+sub print_slaves {
```
Comment (ylacancellera) [inline bin/pt-replica-find]: Since many internal variables were renamed in lib, I'm commenting to be thorough
This tool package did not receive the s/slave/replica
However I don't see any misusage

Unless this was a goal to remove all of them in this PR specifically, I'm ok to keep it and maybe flag it for later

File: bin/pt-replica-find [hunk 9]
```diff
@@ -0,0 +1,4819 @@
...
+   push @lines, ['Uptime', "$dhms (started $started)"];
+
+   my $n_slaves = 0;
+   if ( $proc ) {
+      map { $n_slaves++ if $proc->{$_}->{command} =~ m/Binlog Dump/i }
+         keys %$proc;
+   }
+   push @lines, [
+      'Replication',
+      ($ss ? "Is a slave, " : "Is not a slave, ")
```
Comment (ylacancellera) [inline bin/pt-replica-find]: Unaligned with rest, I can flag it for later if needed

File: bin/pt-replica-find [hunk 10]
```diff
@@ -0,0 +1,4819 @@
...
+  Filters
+  Binary logging  STATEMENT
+  Slave status
+  Slave mode      STRICT
+  Auto-increment  increment 1, offset 1
+  +- 127.0.0.1:12346
+     Version         5.1.34-log
+     Server ID       12346
+     Uptime          04:54 (started 2010-06-17T11:21:24)
+     Replication     Is a slave, has 1 slaves connected
```
Comment (ylacancellera) [inline bin/pt-replica-find]: Unaligned with the rest, but coherent with the current states of things
Again, I can flag it for later, or skip

**Review by ylacancellera — APPROVED** [review]
I have commented what I found already, it's only nitpicks on some missed "slave" wordings. (except the version comparisons, but PT-2390 have been created so I exclude it).

Looks good to me

File: bin/pt-archiver [hunk 11]
```diff
@@ -4161,117 +4205,157 @@ sub wait_for_master {
...
+   my $replica1_name = 'replica';
+   if ( $vp1 lt '8.1' || $vp1->flavor() =~ m/maria/ ) {
+      $replica1_name = 'slave';
+   }
+
+   my $vp2 = VersionParser->new($s2_dbh);
+   my $replica2_name = 'replica';
+   if ( $vp2 lt '8.1' || $vp2->flavor() =~ m/maria/ ) {
+      $replica2_name = 'slave';
+   }
```
Comment (EvgeniyPatlan) [inline bin/pt-archiver]: Maybe we can rework this?

```
sub get_replica_name {
    my ($dbh) = @_;
    my $vp = VersionParser->new($dbh);
    return ($vp lt '8.1' || $vp->flavor() =~ m/maria/) ? 'slave' : 'replica';
}

my $replica1_name = get_replica_name($s1_dbh);
my $replica2_name = get_replica_name($s2_dbh);
```

File: bin/pt-heartbeat [hunk 12]
```diff
@@ -622,117 +666,157 @@ sub wait_for_master {
...
+   if ( $vp1 lt '8.1' || $vp1->flavor() =~ m/maria/ ) {
+      $replica1_name = 'slave';
+   }
+
+   my $vp2 = VersionParser->new($s2_dbh);
+   my $replica2_name = 'replica';
+   if ( $vp2 lt '8.1' || $vp2->flavor() =~ m/maria/ ) {
+      $replica2_name = 'slave';
+   }
+
```
Comment (EvgeniyPatlan) [inline bin/pt-heartbeat]:
```
sub get_replica_name {
    my ($dbh) = @_;
    my $vp = VersionParser->new($dbh);
    return ($vp lt '8.1' || $vp->flavor() =~ m/maria/) ? 'slave' : 'replica';
}

my $replica1_name = get_replica_name($s1_dbh);
my $replica2_name = get_replica_name($s2_dbh);
```

File: bin/pt-kill [hunk 13]
```diff
@@ -4438,117 +4482,157 @@ sub wait_for_master {
...
+   if ( $vp1 lt '8.1' || $vp1->flavor() =~ m/maria/ ) {
+      $replica1_name = 'slave';
+   }
+
+   my $vp2 = VersionParser->new($s2_dbh);
+   my $replica2_name = 'replica';
+   if ( $vp2 lt '8.1' || $vp2->flavor() =~ m/maria/ ) {
+      $replica2_name = 'slave';
+   }
+
```
Comment (EvgeniyPatlan) [inline bin/pt-kill]:
```
sub get_replica_name {
    my ($dbh) = @_;
    my $vp = VersionParser->new($dbh);
    return ($vp lt '8.1' || $vp->flavor() =~ m/maria/) ? 'slave' : 'replica';
}

my $replica1_name = get_replica_name($s1_dbh);
my $replica2_name = get_replica_name($s2_dbh);
```

File: bin/pt-query-digest [hunk 14]
```diff
@@ -11032,117 +11076,157 @@ sub wait_for_master {
...
+   my $replica1_name = 'replica';
+   if ( $vp1 lt '8.1' || $vp1->flavor() =~ m/maria/ ) {
+      $replica1_name = 'slave';
+   }
+
+   my $vp2 = VersionParser->new($s2_dbh);
+   my $replica2_name = 'replica';
+   if ( $vp2 lt '8.1' || $vp2->flavor() =~ m/maria/ ) {
+      $replica2_name = 'slave';
+   }
```
Comment (EvgeniyPatlan) [inline bin/pt-query-digest]:
```

sub get_replica_name {
    my ($dbh) = @_;
    my $vp = VersionParser->new($dbh);
    return ($vp lt '8.1' || $vp->flavor() =~ m/maria/) ? 'slave' : 'replica';
}

my $replica1_name = get_replica_name($s1_dbh);
my $replica2_name = get_replica_name($s2_dbh);

```

**Review by EvgeniyPatlan — APPROVED** [review]

File: lib/PerconaTest.pm (hunk 6 above)
Comment (svetasmirnova) [reply lib/PerconaTest.pm:91]: Will ignore it for now, fix later.

File: bin/pt-archiver (hunk 3 above)
Comment (svetasmirnova) [reply bin/pt-archiver]: Fixed

File: bin/pt-replica-find (hunk 7 above)
Comment (svetasmirnova) [reply bin/pt-replica-find]: Fixed

File: bin/pt-replica-find (hunk 8 above)
Comment (svetasmirnova) [reply bin/pt-replica-find]: fixed

File: bin/pt-replica-find (hunk 10 above)
Comment (svetasmirnova) [reply bin/pt-replica-find]: Fixed

File: bin/pt-replica-find (hunk 9 above)
Comment (svetasmirnova) [reply bin/pt-replica-find]: Fixed

File: lib/MySQLProtocolParser.pm (hunk 5 above)
Comment (svetasmirnova) [reply lib/MySQLProtocolParser.pm]: Changed back to COM_REGISTER_SLAVE , because this is still the name of the constant in 8.4.

File: bin/pt-archiver (hunk 11 above)
Comment (svetasmirnova) [reply bin/pt-archiver]: Implemented

File: bin/pt-kill (hunk 13 above)
Comment (svetasmirnova) [reply bin/pt-kill]: Implemented

File: bin/pt-query-digest (hunk 14 above)
Comment (svetasmirnova) [reply bin/pt-query-digest]: Implemented

**Comment by chaleic** [comment]
@svetasmirnova , now that mysql 8.4 has been merged into the codebase, when might the next version of percona tools be built and the rpm files generated?   Thank you in advance

**Comment by svetasmirnova** [comment]
@chaleic release is in progress, we actively working on it. Please follow Percona's website and blog for updates.
